core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java

```diff
@@ -161,65 +163,79 @@ private Object invoke(ServiceRequestContext ctx, HttpRequest req, @Nullable Aggr
         requireNonNull(pathParams, "pathParams");
         boolean hasRequestMessage = false;
         final ImmutableList.Builder<Parameter> entries = ImmutableList.builder();
-        for (java.lang.reflect.Parameter p : method.getParameters()) {
-            final Param param = p.getAnnotation(Param.class);
+        for (java.lang.reflect.Parameter parameterInfo : method.getParameters()) {
+            final Param param = parameterInfo.getAnnotation(Param.class);
             if (param != null) {
-                final Default aDefault = p.getAnnotation(Default.class);
-                final boolean isOptionalType = p.getType().isAssignableFrom(Optional.class);
-                final boolean isMandatory = !isOptionalType && aDefault == null;
-
-                if (pathParams.contains(param.value())) {
-                    // Path variable
-                    if (!isMandatory) {
-                        throw new IllegalArgumentException(
-                                "Path variable '" + param.value() + "' should not be an optional.");
-                    }
-                    entries.add(Parameter.ofPathParam(validateAndNormalizeSupportedType(p.getType()),
-                                                      param.value()));
-                } else {
-                    // Set the default value to null if it was not specified.
-                    final String defaultValue = aDefault != null ? getSpecifiedValue(aDefault.value()).get()
-                                                                 : null;
-                    final Class<?> type;
-                    if (isOptionalType) {
-                        try {
-                            type = (Class<?>) ((ParameterizedType) p.getParameterizedType())
-                                    .getActualTypeArguments()[0];
-                        } catch (Exception e) {
-                            throw new IllegalArgumentException("Invalid optional parameter: " + p.getName(), e);
-                        }
-                    } else {
-                        type = p.getType();
-                    }
+                entries.add(createParamOrHeaderParam(parameterInfo, ParameterType.PARAM,
```

Comment:
nit: How about moving out the code for making a path variable `Parameter` from `createParamOrHeaderParam` method so that we can let the method make an optional supported `Parameter` only? There may be some code duplication, but it would be more understandable. WDYT?

```
if (pathParams.contains(param.value())) {
    if (parameterInfo.getAnnotation(Default.class) != null ||
        parameterInfo.getType().isAssignableFrom(Optional.class)) {
        throw new IllegalArgumentException(
                "Path variable '" + param.value() + "' should not be an optional.");
    }
    entries.add(Parameter.ofPathParam(
            validateAndNormalizeSupportedType(parameterInfo.getType()), param.value()));
} else {
    entries.add(createParameter(parameterInfo, ParameterType.PARAM, param.value()));
}
continue;
```

Reply:
That is a good idea~! Changed it.